staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go

```diff
@@ -489,7 +490,9 @@ type DeleteOptions struct {
 	// Acceptable values are: 'Orphan' - orphan the dependents; 'Background' -
 	// allow the garbage collector to delete the dependents in the background;
 	// 'Foreground' - a cascading policy that deletes all dependents in the
-	// foreground.
+	// foreground. Note that 'Foreground' deletion is not blocking, and that a
+	// 'Blocking' policy is not available. Even with the 'Foreground' policy,
```

Comment:
I don't understand what "Blocking" means in this context. Blocking what?

Reply:
"Blocking" would block the return from the API call and not send a response until the deletion is complete. [Wikipedia says](https://en.wikipedia.org/wiki/Blocking_%28computing%29):

> A process that is blocked is one that is waiting for some event, such as a resource becoming available or the completion of an I/O operation.

I'm trying to talk about whether a call into the (auto-generated) client API library will wait until the delete operation is done ("blocking", not available), or not ("non-blocking", all of the actually-implemented policies). That's what "Even with the 'Foreground' policy, the resource may still exist after the deletion request succeeds." is trying to convey.

Do you think there's a better way to articulate this?

Reply:
I don't believe any delete calls in our system do that, so it is strange to document it specifically for Foreground deletion.

Reply:
I'm calling out "foreground" specifically because of the name. If I make any sort of call to anything with "foreground" semantics, I expect not to get control back until the whole operation is complete. Here "foreground" is being used to describe the order in which things are deleted (bottom-up), and not whether the process as a whole runs in the foreground or the background from the perspective of the user programming against the API bindings. So I thought it was important to point that out to prevent confusion.

I could drop the part about Foreground and just leave the part about how the resource can exist after the request succeeds, if you think that would be better.

Reply:
Is the python documentation presenting this in a context that makes it sound like it's about the api call mechanics rather than the garbage collector behavior? @roycaihw do you have thoughts?

Reply:
The [python documentation](https://github.com/kubernetes-client/python/blob/7fb3a25176395da26007aec7b4c843aa03514c21/kubernetes/docs/BatchV1Api.md#parameters-2) uses the same description from the openapi spec:

Name | Type | Description  | Notes
------------- | ------------- | ------------- | -------------
 **propagation_policy** | **str**| Whether and how garbage collection will be performed. Either this field or OrphanDependents may be set, but not both. The default policy is decided by the existing finalizer set in the metadata.finalizers and the resource-specific default policy. Acceptable values are: &#39;Orphan&#39; - orphan the dependents; &#39;Background&#39; - allow the garbage collector to delete the dependents in the background; &#39;Foreground&#39; - a cascading policy that deletes all dependents in the foreground. | [optional] 

and I didn't find the context suggesting any capability of blocking API calls.

Reply:
I'd start [a bit higher on the page](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/BatchV1Api.md#delete_namespaced_job), where the description of what the `delete_namespaced_job` method does says that it will:
> delete a Job

This short, simple, and clear description of a Python method gives no hint of the possibility that the method might return without actually having deleted the job. It doesn't suggest any capability of *non*-blocking Python method calls.

The documentation then gives a piece of example code (which, incidentally, cannot possibly work, as many of the string arguments are outside the set of acceptable values), and the table of argument semantics.

For `grace_period_seconds`, it says that "The value zero indicates delete immediately.", which suggests that an immediately-visible deletion is at least possible. If you read and think about the description you realize that operations are then possibly not immediate by default, if the "per object value" for Job is not zero. (Where would one actually find that per object value, by the way?) However, even that doesn't contradict synchronous semantics, especially if you interpret this as being passed along as the grace period given to the job's pods after they are asked to terminate. They could shut themselves down in a few tenths of a second, allowing a whole synchronous delete operation to succeed quickly.

The `propagation_policy` description refers to "garbage collection" that happens as a result of my request, and not the operation I am directly performing. My reading in this context is that `Background` will let the call return after the job itself is gone, with Kubernetes cleaning up any pods in the background, while `Foreground` will wait for the pod to go away before the job deletion completes, and thus before the next statement in my Python code runs.

The fact that the ability to put some of the cleanup of dependent operations into the background is described in such loving detail, while no mention is made of the synchronicity or asynchronicity of the call as a whole, leads the application developer to conclude that they are looking at a synchronous operation, or at least an operation that has been made synchronous from their perspective by lower-level code. If they ask for `Foreground` deletion with a 0 grace period, they would be right, or at least not wrong, to expect an immediate and immediately visible result.

Overall, it's pretty clear that the Python API calls map 1:1 with web requests, but the Python documentation doesn't give any hint that the semantics of those web requests are that they will immediately return success before the operation being requested has occurred, and that there's no mechanism in the Python wrapper to protect applications from this behavior. This principle of non-blocking, asynchronous operation everywhere is not mentioned in [the main page of the Python API docs](https://github.com/kubernetes-client/python/tree/7fb3a25176395da26007aec7b4c843aa03514c21/kubernetes) or in the [Python project README](https://github.com/kubernetes-client/python/blob/7fb3a25176395da26007aec7b4c843aa03514c21/README.md), even in the Troubleshooting section. People reading the Python documentation probably haven't also already read the introductory documentation for the Kubernetes web API itself; generally you learn by going from `kubectl` straight to Python, and I've never been quick enough to catch the return-before-complete with `kubectl`. I routinely say `kubectl delete job blah; kubectl apply -f blah-job.yml` and have never once had that construction try to patch an existing job.

I may have talked myself out of the propagation policy descriptions as being quite the right place to explain that, overall, the Python bindings don't protect the application and guarantee read-your-own-writes consistency in the same way that kubectl protects and guarantees this to the user. How do y'all think that might be best communicated?

Reply:
`kubectl delete` waits for resources to be gone before returning by default. The python client behaves the same as the other clients (client-go, client-java):
```
$ kubectl delete -h
      --wait=true: If true, wait for resources to be gone before returning. This waits for finalizers.
```

Maybe we can add explanation to the python client README / an API concept doc that captures FAQs in the python client.

Reply:
I do think this is a common difference from kubectl that surprises api users, even in the go client. Maybe a simple change that would help is to change "delete a job" to "Begin deleting a job." Ideally, we'd change this for every language, if possible.